Support Yarn backward compatibility mode (node-modules) for Yarn version 2.x or higher

nenadvicentic opened the pull request:
This pull request addresses issue #85.

When development machine has `yarn` version 2.x or higher installed (current version at the moment is 3.x), after developer chooses to install `npm` dependencies using `yarn`, `yarn` works in [Plug'n'Play](https://yarnpkg.com/features/pnp) mode. This means that no `node_modules` folder in created and all `npm` packages are stored in `.yarn` folder unpacked, in original `zip` format.

1. This means that command `npm start` written in "Getting started" tip does not work. `yarn start` is printed now, when `yarn` was picked as package manager. 
2. ~~When using Visual Studio Code, for it's TypeScript language service to work Yarn's [Editor SDKs](https://yarnpkg.com/getting-started/editor-sdks) extension is needed, to make TypeScript language service (local to the folder) aware of module resolution strategy.~~
3. ~~For Parcel not to throw errors, two dev-dependencies had to be added to project's `package.json`: "@aurelia/runtime-html", "@aurelia/router" for direct-routing template and "@parcel/config-default".~~

~~Now parcel starts/builds project without throwing errors.~~

**UPDATE**:
It turned out to be a problem to make Yarn's new PnP mode work seamlessly with different IDEs and bundlers. Switching Yarn 2+ to backward compatibility mode with `nodeLinker: node-modules` option, to force it to create `node_modules` folder seems like more stable and simpler option at the moment.

3cp commented:
Thx very much for working on this!
The parcel deps issue is bit strange. Probably parcel+yarn-pnp issue. I will test more and get back to you.

3cp commented on after.js:
```diff
@@ -35,17 +40,37 @@ module.exports = async function({
+        if (packageManager === 'npm') {
+          await run('npx', ['playwright', 'install', '--with-deps']);
+        } else {
+          await run(packageManager, ['dlx', 'playwright', 'install', '--with-deps']);
```
I assume this is only for "yarn dlx".
For pnpm, it's pnpx.

nenadvicentic replied:
Logic is to execute `pnpm dlx` and `yarn dlx` depending on a name chosen.
Documentation for `pnpm dlx` is [here](https://pnpm.io/cli/dlx).  I did test this path.

3cp replied:
Thx! I had no idea what is dlx.

3cp commented on parcel/package.json:
```diff
@@ -1,7 +1,12 @@
+    "@aurelia/router": /* @if !dev */"latest"/* @endif *//* @if dev */"dev"/* @endif */,
+    /* @endif */
     "@aurelia/parcel-transformer": /* @if !dev */"latest"/* @endif *//* @if dev */"dev"/* @endif */,
+    "@parcel/config-default": "^2.6.0",
```
https://github.com/parcel-bundler/parcel/issues/7999

I tested yarn pnp loose mode. It didn't help.
I really dislike this patch for parcel's issue.

Instead of doing here which unnecesarily affect users don't use yarn pnp, we can conditionally do "yarn add the-three" in after.js like you did for vscode integration.

nenadvicentic replied:
Since you commented that this also fixes webpack [here](https://github.com/aurelia/new/issues/85#issuecomment-1260225520), should I move those 2 dependencies in `common/package.json`?

3cp replied:
Yarn add will update package.json. But this will only update package.json for users who do uses pnp.

You can test the same approach for webpack with pnp, I have not tested that.

3cp replied:
To be honest, I am quite confused in the whole parcel and yarn pnp github issues. I am not interested to dig deep . Sorry :-) I had enough experience with yarn and pnpm.

I used yarn and pnpm both for quite some time few years ago, and contributed to them both. But finally I moved back to npm because I really don't want to spend my time to understand the difference anymore. The little saving on install time didn't justify the troubles (both known and unknown) I faced.

nenadvicentic replied:
@3cp The problem is - default skeletons could be (and could have been) one of the main reasons for poor `aurelia` adaption. I'm using `aurelia` since 2015 and there were always problems. If 30-50% of combinations for scaffolding do not work, that's not good.

Somewhere during my investigation I read that rule-of-thumb should be - you don't explicitly reference a module in your code which is not explicitly in your `package.json` dependencies. Possibly even linter rule... This makes sense to avoid error and applications unexpectedly breaking after update of packages.

In general, if `npm` was good enough, there would not be `yarn` and `pnpm`. If I'm not wrong they are now both shipped with `node` installation. Handled via [Corepack](https://nodejs.org/dist/latest/docs/api/corepack.html). So people will eventually move from `yarn` v1 (current is 3.2.3) and they will immediately run into all of those issues, same way I did.

Additional note - even after we solve the issue of adding extra packages, Parcel cannot do tree-shaking because of dynamically added module `@aurelia/runtime-html`. Maybe this modification should be done earlier in the pipeline, so that tree-shaking works. I mentioned this already, and you can see warning if you add `--trace-warnings` flag to `parcel` command(s).

Regarding `yarn add` in `after.js`, it's clear. I will do the change.

3cp replied:
One of the main troubles of au skeletons is due to we support too many combinations. Some of them like to introduce breaking changes very often :-)

Also I believe skeletons from many other communities assume npm in use.

I was also trying to reduce au2's skeleton options deliberately (comparing to au1). :-)

nenadvicentic replied:
@3cp opinion needed.

> You can test the same approach for webpack with pnp, I have not tested that.

So, as we discussed, I've been testing if adding additional packages by executing `yarn add` in `after.js` would fix the issues with Webpack. It didn't.

But additional problem appeared. While creating new project from the skeleton, because we, correctly, have in `package.json`:
```json
"typescript": "^4.7.3"
```
The latest stable version was picked up: `4.8.4`. It turned out that this version breaks Yarn's Visual Studio Code extension:
https://github.com/yarnpkg/berry/pull/4903.
The fix for `yarn dls @yarnpkg/sdks vscode` will be released next week, but this prompted me to look what is the root cause and it's Microsoft's refusal (2 years of discussion) to integrate `pnp`  module resolution directly, or to enable plugin infrastructure (for this I think they are right). Here are the details:
https://github.com/microsoft/TypeScript/issues/28289

So, after thinking about this, I have to agree with you, `Plug'n'Play` and `TypeScript` is a tricky business at the moment and things can break or at least have hiccups in the future.

Perhaps it would be safer approach to force `yarn` to create `node_modules`. This can be achieved by adding `.yarnrc.yml` file in the root of the project with the following entry:
>nodeLinker: node-modules

This configuration switch is working with both Parcel and Webpack. It enforces `node_modules` to be created, but at least it won't leave user with a newly created project broken. I also think we should keep `yarn start` and `pnpm start` display messages.

@3cp What do you think about this?

3cp replied:
The npm/yarn/pnpm question only pops up in after.js, and it's behind the check of silent mode.

Since deps install is not a mandatory step for scaffolding a project, this is not in questions.js.

When user ran makes in silent mode, we don't bother to ask user for deps installation. It's totally up to the user what to do next. The silent mode is designed for automation script, where we do as little as required.

The yarn config should work. We can present some help text (question choice has description) to clarify yarn pnp doesn't work well with au2.

This kind of ts/vscode/yarn issue is exactly what I complained about skeleton maintenance:-)

3cp replied:
You don't have to read package.json about ^2.6.0, it means any latest major version 2 at least higher then 2.6.0. It will always install latest parcel v2, not v2.6.0.

Just run yarn add foo, it will put ^latest.foo.version in package.json.

Parcel is not going to v3 for anytime soon.

nenadvicentic replied:
> The npm/yarn/pnpm question only pops up in after.js, and it's behind the check of silent mode.

You can simply default to `npm` in silent mode. Moving it to `questions.js` would allow to have files:
`yarn/.yarnrc.js` or `pnpm/.npmrc`, instead of  everything in `common` folder.

It would also allow, for example, in `typescript/package.json`:

```json
{
  "devDependencies": {
    // @if typescript && yarn-pnp
    "@parcel/config-default": "^2.6.0"
    // @endif
   "parcel": "^2.6.0"
```
So, all versions in `json` instead of `after.js`:
```js
await run('yarn', ["add", "@parcel/config-default@^2.6.0", "--dev"]);
```
Currently you have to update version `^2.6.0` in two different (non-intuitive) places.

However, this 2nd example is theoretical, because once we add `.yarnrc.yml` with `nodeLinker: node-modules`, we can remove handling of extra dependencies in `after.js` for Yarn's Plug'n'Play mode.

Moving the package-manager question might have pros and cons, so I am not arguing it's better. Just trying to make it clear.

> The yarn config should work. We can present some help text (question choice has description) to clarify yarn pnp doesn't work well with au2.

I had in mind to something along these lines:
```shell
npm
yarn (classic)
pnpm
```

3cp replied:
I mean we cannot assume user want to do npm install in silent mode. There is usage of private npm registry, user might want to touch .npmrc before deps install.

When design tool, I aid to have less assumption, that enables flexible use cases.

Putting them in question.js also means that's in our offical support. We have to cover yarn and pnpm in CI setup. But we are definitely not keen to officially support future versions of yarn and pnpm.

3cp replied:
Wait a min.
With `nodeLinker: node-modules`, is not that means we don't need to explicit install `@parcel/config-default`?

3cp replied:
Yes, I confirmed `nodeLinker: node-modules` means it all worked out with explicitly adding of deps.

3cp replied:
Why not :)
Our pnpm setup also uses patch in npmrc config yoo.

3cp commented on __test__/after-task.spec.js:
```diff
@@ -79,7 +79,7 @@ test('"after" task only prints summary in unattended mode and here mode', async
 });
 
-test('"after" task installs deps, and prints summary', async t => {
+test('"after" task installs deps with yarn, and prints summary', async t => {
```
Please duplicate this test case, one to cover npm, one for yarn.

3cp commented:
Unfortunately yarn pnp loose mode didn't help on the parcel issue :-(

nenadvicentic commented:
@3cp After thinking a bit longer about this, I reworked the branch from scratch:

1. Instead of trying to make everything work with Yarn's PnP, I switched it to backward compatible mode with `nodeLinker: node-modules` option in `.yarnrc.yml` file. Now Yarn 2+ created `node_modules` folder as well.
2. As discussed before "Get Started" displays name of correct package manager. Unit tests are all there.
3. I added `strict-peer-dependencies=false` option for `pnpm` in `.npmrc` file. To avoid installation of packages breaking due to missing peer dependency.

3cp commented:
Thank you!
